File: design/tree.go
package design

// Definition for a binary tree node.
type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func MakeTree(index int, vals []any) *TreeNode {
	if len := len(vals); len == 0 || len <= index || vals[index] == nil {
		return nil
	}
	return &TreeNode{
		Val:   vals[index].(int),
		Left:  MakeTree(2*index+1, vals),
		Right: MakeTree(2*index+2, vals),
	}
}

func MakeTree2(vals ...any) *TreeNode {
	var root *TreeNode = nil
	var queue = make([]*TreeNode, 0)
	if len(vals) > 0 && vals[0] != nil {
		root = &TreeNode{Val: vals[0].(int)}
		queue = append(queue, root)
	}
	var curr *TreeNode = nil
	for idx, n := 1, len(vals); idx < n; idx++ {
		var child *TreeNode = nil
		if vals[idx] != nil {
			child = &TreeNode{Val: vals[idx].(int)}
			queue = append(queue, child)
		}
		if curr == nil {
			curr = queue[0]
			queue = queue[1:]
			curr.Left = child
		} else {
			curr.Right = child
			curr = nil
		}
	}
	return root
}

func traversalBFS(root *TreeNode) []int {
	if root == nil {
		return []int{}
	}
	var values = make([]int, 0)
	var queue = NewQueue[*TreeNode]()
	queue.Enqueue(root)
	for queue.Size() > 0 {
		current := queue.Dequeue()
		values = append(values, current.Val)
		if current.Left != nil {
			queue.Enqueue(current.Left)
		}
		if current.Right != nil {
			queue.Enqueue(current.Right)
		}
	}
	return values
}

func traversalDFSInOrder(root *TreeNode) []int {
	if root == nil {
		return []int{}
	}
	var values = make([]int, 0)
	if root.Left != nil {
		values = append(values, traversalDFSInOrder(root.Left)...)
	}
	values = append(values, root.Val)
	if root.Right != nil {
		values = append(values, traversalDFSInOrder(root.Right)...)
	}
	return values
}

func traversalMorris(root *TreeNode) []int {
	var values = make([]int, 0)
	var prev *TreeNode
	var curr = root
	for curr != nil {
		if curr.Left == nil {
			values = append(values, curr.Val)
			curr = curr.Right
		} else {
			prev = curr.Left
			for prev.Right != nil && prev.Right != curr {
				prev = prev.Right
			}
			if prev.Right == curr {
				prev.Right = nil
				values = append(values, curr.Val)
				curr = curr.Right
			} else {
				prev.Right = curr
				curr = curr.Left
			}
		}
	}
	return values
}

// Definition for a n-ary tree node.
type TNode struct {
	Val      int
	Children []*TNode
}

func MakeNAryTree(vals []any) *TNode {
	if len(vals) == 0 || vals[0] == nil {
		return nil
	}
	var root = &TNode{Val: vals[0].(int), Children: make([]*TNode, 0)}
	var parents = []*TNode{root}
	var index = 2
	for length := len(parents); length > 0; length = len(parents) {
		for i := 0; i < len(parents); i++ {
			for index < len(vals) && vals[index] != nil {
				var child = &TNode{
					Val:      vals[index].(int),
					Children: make([]*TNode, 0),
				}
				parents[i].Children = append(parents[i].Children, child)
				parents = append(parents, child)
				index++
			}
			index++
		}
		parents = parents[length:]
	}
	return root
}

File: pkg/tree.py
from collections import deque
from typing import List, Optional, Any


# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val: int = 0, left: "TreeNode" = None, right: "TreeNode" = None):
        self.val = val
        self.left = left
        self.right = right

    def __str__(self) -> str:
        if not self:
            return "None"
        return f"{self.val},{self.left.__str__()},{self.right.__str__()}"

    def __eq__(self, __value: object) -> bool:
        if not __value:
            return False
        return all(
            [
                self.val == __value.val,
                (not self.left and not __value.left or self.left and self.left.__eq__(__value.left)),
                (not self.right and not __value.right or self.right and self.right.__eq__(__value.right)),
            ]
        )


# def create_tree(vals: List[int], idx: int) -> Optional[TreeNode]:
#     if idx >= len(vals) or vals[idx] == None:
#         return None
#     return TreeNode(
#         val=vals[idx],
#         left=create_tree(vals, 2*idx+1),
#         right=create_tree(vals, 2*idx+2)
#     )


def create_tree(vals: List[Any]) -> Optional[TreeNode]:
    root = None
    queue = deque()
    if len(vals) > 0 and vals[0] is not None:
        root = TreeNode(vals[0])
        queue.append(root)
    curr = None
    for idx in range(1, len(vals)):
        child = None
        if vals[idx] is not None:
            child = TreeNode(vals[idx])
            queue.append(child)
        if not curr:
            curr = queue.popleft()
            curr.left = child
        else:
            curr.right = child
            curr = None
    return root


# Definition for a n-ary tree node.
class Node:
    def __init__(self, val: int = 0, children: List = None):
        self.val = val
        self.children = children or []


def create_n_ary_tree(vals: List[Any]) -> Optional[Node]:
    if len(vals) == 0 or vals[0] is None:
        return None
    root = Node(vals[0])
    queue = deque()
    queue.append(root)
    idx = 2
    while idx < len(vals):
        while queue:
            node = queue.popleft()
            while idx < len(vals) and vals[idx] is not None:
                child = Node(vals[idx])
                node.children.append(child)
                queue.append(child)
                idx += 1
            idx += 1
        idx += 1
    return root
